File: sudoku.py
import numpy as np

def creategame(FILE):
    f = open(FILE, "r")
    game = np.mat(f.read())
    f.close()
    return game

def checkrow(game, row, num):
    for x in range(9):
        if game[row, x] == num:
            return False
    return True

def checkcolumn(game, column, num):
    for x in range(9):
        if game[x, column] == num:
            return False
    return True

def checksquare(game, square, num):
    if square < 3:
        if square == 0:
            if num in game[0:3, 0:3]:
                return False
            return True
        elif square == 1:
            if num in game[0:3, 3:6]:
                return False
            return True
        elif square == 2:
            if num in game[0:3, 6:9]:
                return False
            return True
    elif square < 6:
        if square == 3:
            if num in game[3:6, 0:3]:
                return False
            return True
        elif square == 4:
            if num in game[3:6, 3:6]:
                return False
            return True
        elif square == 5:
            if num in game[3:6, 6:9]:
                return False
            return True
    else:
        if square == 6:
            if num in game[6:9, 0:3]:
                return False
            return True
        if square == 7:
            if num in game[6:9, 3:6]:
                return False
            return True
        if square == 8:
            if num in game[6:9, 6:9]:
                return False
            return True

def whichsquare(row, column):
    if row < 3:
        if column < 3:
            return 0
        elif column < 6:
            return 1
        else:
            return 2
    elif row < 6:
        if column < 3:
            return 3
        elif column < 6:
            return 4
        else:
            return 5
    else:
        if column < 3:
            return 6
        elif column < 6:
            return 7
        else:
            return 8

def check(game, row, column, square, num):
    x = checkrow(game, row, num)
    y = checkcolumn(game, column, num)
    z = checksquare(game, square, num)
    if x and y and z:
        return True
    return False

def next(row, column):
    if (column + 1) > 8:
        return (row + 1, 0)
    return (row, column + 1)

def find0(game, coord=(0, 0)):
    row, column = coord
    if game[row, column] != 0:
        return find0(game, next(row, column))
    return coord

def complete(game):
    if 0 in game:
        return False
    return True

def solving(game, row=0, column=0):

    if complete(game):
        return True
    for x in range(1, 10):
        row, column = find0(game, (row, column))
        if check(game, row, column, whichsquare(row, column), x):
            game[row, column] = x
            if solving(game):
                return True
            game[row, column] = 0
    return False

def main():
    sudoku = creategame('Sudoku.txt')
    solving(sudoku)
    print(sudoku)

File: sudoku.c
#include <stdio.h>
#include <stdlib.h>

int **creategame(FILE *cp);
void freegame(int **game);
void printgame(int **game);
int check_row(int **game, int row, int column, int num);
int check_column(int **game,int row, int column, int num);
int check_square(int **game, int row, int column, int num);
int check_num(int **game, int row, int column, int num);
void find_next(int **game, int *x, int *y);
int check_all(int **game);
int backtracking(int **game, int x, int y);

int main() {
	FILE *pfile = fopen("Sudoku.txt","r");
	int **sudoku = creategame(pfile);
	backtracking(sudoku, 0, 0);
	printgame(sudoku);
	freegame(sudoku);
	return 0;
}

int **creategame(FILE *cp) {
	int **game;
	if ( (game = (int**) malloc(9*sizeof(int*))) != NULL) {
		for (int i = 0; i < 9; i++) {
			if ( (game[i] = (int*) malloc(9*sizeof(int))) == NULL) {
				while (i) {
					i--;
					free(game[i]);
				}
				free(game);
				return NULL;
			}
		}
	}
	for (int i = 0; i < 9; i++) {
		for (int j = 0; j < 9; j++) {
			fscanf(cp, "%d", &game[i][j]);
		}
		fseek(cp, 2, SEEK_CUR);
	}
	return game;
}

void freegame(int **game) {
	for (int i = 0; i < 9; i++) {
		free(game[i]);
	}
}

void printgame(int **game) {
	for (int i = 0; i < 9; i++) {
		for (int j = 0; j < 9; j++) {
			printf("%d ", game[i][j]);
		}
		printf("\n");
	}
}

int check_row(int **game, int row, int column, int num) {
	for (int i = 0; i < 9; ++i) {
		if (game[row][i] == num) {
			if (i != column) {
				return 0;
			}
		}
	}
	return 1;
}

int check_column(int **game,int row, int column, int num) {
	for (int i = 0; i < 9; ++i) {
		if (game[i][column] == num) {
			if (i != row) {
				return 0;
			}
		}
	}
	return 1;
}

int check_square(int **game, int row, int column, int num) {
	int i, j, p, q;
	if (row <= 2) {
		i = 0; j = 2;
	}
	else if (row <= 5) {
		i = 3; j = 5;
	}
	else {
		i = 6; j = 8;
	}
	if (column <= 2) {
		p = 0; q = 2;
	}
	else if (column <= 5) {
		p = 3; q = 5;
	}
	else {
		p = 6; q = 8;
	}
	for (int b = i; b <= j; b++) {
		for (int n = p; n <= q; n++) {
			if (game[b][n] == num) {
				if (b != row || n != column) {
					return 0;
				}
			}
		}
	}
	return 1;
}

int check_num(int **game, int row, int column, int num) {
	int flag1, flag2, flag3;
	flag1 = check_row(game, row, column, num);
	flag2 = check_column(game, row, column, num);
	flag3 = check_square(game, row, column, num);
	if (flag1 == 1 && flag2 == 1 && flag3 == 1) {
		return 1;
	}
	return 0;
}

void find_next(int **game, int *x, int *y) {
	while (game[*x][*y] != 0) {
		if ((*y) != 8) { (*y)++; }
		else { (*x)++; (*y) = 0; }
	}
}

int check_all(int **game) {
	for (int i = 0; i < 9; i++) {
		for (int j = 0; j < 9; j++) {
			if (game[i][j] == 0) {
				return 0;
			}
		}
	}
	return 1;
}

int backtracking(int **game, int x, int y) {
	if (check_all(game)) {
		return 1;
	}
	for (int test = 1; test <= 9; test++) {
		find_next(game, &x, &y);
		if (check_num(game, x, y, test)) {
			game[x][y] = test;
			if (backtracking(game, x, y)) {
				return 1;
			}
			game[x][y] = 0;
		}
	}
	return 0;
}
